File: vol2/ch17/LeetCode410.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        low,high=max(nums),sum(nums)
        while low<high:
            mid=(low+high)//2
            cnt=self.Count(nums,mid)
            if cnt<k:           #说明mid大了
                high=mid       #在左区间中继续查找
            else:                #说明mid小了
                low=mid+1     #在右区间中继续查找
        return low
    def Count(self,nums,mid):
        cursum,cnt=0,0
        for i in range(0,len(nums)):   #求以mid为运载能力时对应的天数cnt
            cursum+=nums[i]
            if cursum>mid:
                cnt+=1
                cursum=nums[i]
        return cnt

File: vol2/ch17/LeetCode410.cpp
class Solution {
public:
    int splitArray(vector<int>& nums,int m) {
        int n=nums.size();
        int low=0,high=0;
        for(int e:nums) {
            low=max(low,e);
            high+=e;
        }
        while (low<high) {
            int mid=(low+high)/2;
            int cnt=Count(nums,mid);
            if (cnt<m)              //说明mid大了
                high=mid;           //在左区间中继续查找
            else                    //说明mid小了
                low=mid+1;          //在右区间中继续查找
        }
        return  low;
    }
    int Count(vector<int>& nums, int mid) {
        int cursum=0,cnt=0;
        for (int i=0;i<nums.size();i++) {
            cursum+=nums[i];
            if (cursum>mid) {
                cursum=nums[i];
                cnt++;
            }
        }
        return cnt;
    }
};

